src/sqlfluff/core/rules/base.py

```diff
@@ -97,7 +110,7 @@ def __init__(
         self,
         anchor: Optional[BaseSegment] = None,
         fixes: Optional[List["LintFix"]] = None,
-        memory=None,
+        memory: Optional[Iterable[Any]] = None,
```

Comment:
This is a tricky one. The original _intent_ of `memory` was that it was whatever data structure a given rule wanted it to be. That isn't very helpful for mypy though. Does `Optional[Any]` work here?